### ISSUE TITLE
[globus-connect-personal] $USER discovery

### ISSUE THREAD

**Issue by vsoch** [globus/globus-cli#394]
This is an issue for the globusconnectpersonal client, I couldn't find where to post it so I hope it's ok to post here! Please redirect if necessary, the download of the client is from amazon s3 so I'm not sure if the code is public.

-------------------------------------------------------------------------------------------------
I am setting up an endpoint, and my environment doesn't have `$USER` defined. It leads to this:
```bash
$ /opt/globus/globusconnectpersonal -setup "${toke 
Traceback (most recent call last):
  File "/opt/globus/gc-ctrl.py", line 61, in <module>
    os.environ['GCP_USER'] = os.environ['USER']
  File "/usr/lib/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'USER'
globus-user@c0138b646b51:/code$ echo $USER

```

Ahh! So that also shows the script that needs help, `gc-ctrl.py` is extracted from the latest [referenced here](https://docs.globus.org/how-to/globus-connect-personal-linux/#globus-connect-personal-cli).

I think instead of relying on the environment variable, we should use the module `pwd` to get the userid and look it up from that. Here is how I would do it in the same environment (where user isn't defined)

```python
import os
import pwd

os.getuid()
1000

# This returns an object with lots of goodies!
pwd.getpwuid(os.getuid())
pwd.struct_passwd(pw_name='globus-user', pw_passwd='x', pw_uid=1000, pw_gid=1000, pw_gecos='', pw_dir='/home/globus-user', pw_shell='/bin/bash')

# Here is the $HOME 
pwd.getpwuid(os.getuid())[5]
'/home/globus-user'

# And here is the user!
pwd.getpwuid(os.getuid())[0]
'globus-user'
```
So I think this bug would be fixed with:

```python
import os
import pwd
os.environ['GCP_USER'] = pwd.getpwuid(os.getuid())[0]
```

**Comment by vsoch** [comment]
Because it gets the effective user, not the one that happens to be exported in the environment, in case that matters.

**Comment by vsoch** [comment]
This is kind of interesting, at some point `$USER` gets unset...
```
$ which globus
/opt/conda/bin/globus
$ whoami
tunel-user
$ globus login --no-local-server
Please log into Globus here:
---------------------------
<removed>
---------------------------

Enter the resulting Authorization Code here: <removed>

You have successfully logged in to the Globus CLI as <removed>

You can always check your current identity with
  globus whoami

Logout of the Globus CLI with
  globus logout

response=$(globus endpoint create --personal "${ROBOTNAME}")
$ for token in ${response}; do token=$token;done
$ echo $token
<removed>
$ echo $USER

$ export USER=$(whoami)
$ echo $USER
tunel-user
```
mysterious!!

**Comment by sirosen** [comment]
There is no public repository for globus connect personal, but even if some of the source remains closed, maybe there should be...
There's not much to be gained by rerouting this issue right now. Fixing the problem comes first. But maybe we'll make some changes for reporting bugs. Otherwise, I don't mind these sorts of things being reported here, via support@globus.org, or on the developer-discuss@globus.org listhost.

It's unlikely that the Globus CLI is having an impact, since it executes in a subprocess and therefore cannot change the environment of the calling shell.

I'll make sure the Globus Connect Personal maintainers see it. My initial reaction is that you're right and we should be detecting the current user in a more robust manner. `pwd` from the standard lib sounds like a good option -- possibly just as a failover when `USER` is unset.

**Comment by vsoch** [comment]
okay sounds good. I generally try to report based on finding the right code base, so I'm +1 for having some of the code exposed, or even a mostly empty repository named appropriately so developers know "that's the one!" 

Thanks for forwarding the issue! If the developers want, we can discuss here, or if there is a better place close here and I can repost there. It's not a direly urgent thing - nothing is broken, it's more of edge case I think (although I hit it, so maybe not so unlikely!)

**Comment by sirosen** [comment]
We have a patch which uses `getpass.getuser()` undergoing review right now. It's a slightly higher-level stdlib function which checks several env vars, including `$USER`, and then fails over to `pwd`.
I'll update here when the fix is released.

---

As an aside, I notice that you're parsing the textual output of `globus endpoint create --personal`

This is dangerous -- we don't guarantee the stability and parsability of `--format text` output. In practice we're well aware that people do it and try to keep output stable.
Something line-oriented like `awk '/Setup Key:/{print $3}'` is incrementally safer, but I wouldn't recommend that either.

The best thing is to extract values from the JSON output. For example, with `--jmespath` you can do this:
```
$ globus endpoint create --personal "tmpep" --jmespath 'globus_connect_setup_key'
"e0c414bc-ed4b-4986-a2de-22d68c998c1f"
```

Note that the output string will be quoted, so piping `tr -d '"'` may be handy too.

**Comment by vsoch** [comment]
That worked great!
```
$ token=$(globus endpoint create --personal "${ROBOTNAME}" --jmespath 'globus_connect_setup_key'  | tr -d '"') 
$ echo $token
ef3fedb1-a93d-4053-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
Thank you for the help!

**Comment by sirosen** [comment]
Glad to see that the `--jmespath` invocation works for you!
I recommend always using `--jq/--jmespath` when you need to parse CLI output, or otherwise piping full JSON output to tools which can parse it. Refactoring our documentation to strongly encourage its use is on my ToDo list.

The Globus Connect Personal team just notified me today that they've released Globus Connect Personal Linux v2.3.5, which switches to use `getpass.getuser`.
I'm going to close this, but feel free to reopen if you still see issues with the latest version.
